Adding new topic for gas in Sui (#10245)
MystenLabs/sui: doc/src/learn/tokenomics/gas-budgets.md

---

> title: Sui Gas Budgets

Reviewer: ```suggestion
title: Gas in Sui
```

---

> Gas Budgets are important because they guarantee that an user never pays more transaction fees than what they budget for. Sui’s gas mechanism is such that users pay upfront for both the current computational cost of executing a transaction and the long-term cost of storing the data associated with a transaction’s objects. Thus, Gas Budgets need to cover both sets of fees.

Reviewer: ```suggestion
A Sui transaction must pay for both the computational cost of execution and the long-term cost of storing the objects a transaction creates or mutates.
```

Reviewer: Since this is so much shorter, I would suggest moving line 7 into this paragraph.

> Specifically, [Sui’s Gas Pricing Mechanism](https://docs.sui.io/learn/tokenomics/gas-pricing) is such that any transaction $\tau$ pays the following gas fees:

Reviewer: ```suggestion
Specifically, [Sui’s Gas Pricing Mechanism](https://docs.sui.io/learn/tokenomics/gas-pricing) is such that any transaction pays the following gas fees:
`total_gas_fees = computation_units * reference_gas_price + storage_units * storage_price`
```

$$
\text{GasFees}[\tau] = \text{ComputationUnits}[\tau]\times \text{ReferenceGasPrice}\ + \ \text{StorageUnits}[\tau]\times \text{StoragePrice}
$$

While computation and storage fees are separate, they are conceptually similar in that they each translate computation or storage into SUI terms by multiplying computation or storage units by the relevant price. 

> ### Gas Prices

Reviewer 2: Sentence case on subheadings

The [Reference Gas Price](gas-pricing.md#computation-gas-prices) translates the real-time cost of executing a transaction into SUI units and is updated at each epoch boundary by the validator set. Similarly, the [Storage Price](gas-pricing.md#storage-gas-prices)  translates the long-term cost of storing data on-chain into SUI units and is updated infrequently; often remaining constant for various consecutive epochs. During regular network operations, all Sui users should expect to pay the Reference Gas Price and Storage Price for computation and storage, respectively.

> ### Gas Units

Reviewer: Let's make sure the table from the Notion doc that shows each bucket makes it into here.

Different Sui transactions require varying amounts of computational time in order to be processed and executed. Sui translates these varying operational loads into transaction fees by measuring each transaction in terms of Computation Units. All else equals, more complex transactions will require more Computation Units. 

> Importantly, though, Sui’s gas schedule is built coarsely with a bucketing approach. Two relatively similar transactions will translate into the exact same amount of Computation Units if they are in the same bucket, whereas two relatively different transactions will translate into different amounts of Computation Units if they fall in separate buckets. The smallest bucket maps into 1000 Computation Units, meaning that all transactions that fall into the smallest bucket will cost 1000 Computation Units. The largest bucket maps into 5,000,000 Computation Units; if a transaction were to require more Computation Units it would simply abort.

Reviewer 2: Why capitalize Computation Units and Storage Units?

> 

Reviewer: ```suggestion
Using coarse bucketing accomplishes two important goasl:
- Frees developers from optimizing their smart contracts to deliver marginal gains in gas costs via "gas golfing"--instead, can focus on step-function improvements in their products and services.
- Gives Sui protocol devs the freedom to adjust per-instruction gas costs and experiment with new gas metering schemes without creating significant disruption for builders. We expect this will happen frequently, so it's important that builders do *not* rely on a per-instruction gas costs remaining stable over time.
```

> Similarly, Sui transactions vary depending on the amount of new data written into on-chain storage. The variable Storage Units captures these difference by mapping the amount of bytes held in storage into storage units. Sui’s current schedule is linear and maps each byte into 100 storage units. So, for example, a transaction that stores 25 bytes will cost 2500 Storage Units while a transaction that stores 75 bytes will cost 7500 units.

Reviewer: ```suggestion
###  Storage Units
Sui transactions vary depending on the amount of new data written into on-chain storage. The variable Storage Units captures these difference by mapping the amount of bytes held in storage into storage units. Sui’s current schedule is linear and maps each byte into 100 storage units. So, for example, a transaction that stores 25 bytes will cost 2500 Storage Units while a transaction that stores 75 bytes will cost 7500 units.
```

### Gas Budgets

All transactions need to be submitted together with a Gas Budget. This provides a cap to the amount of Gas Fees a user will pay, especially since in some cases it may be hard to perfectly forecast how much a transaction will cost before it is submitted to the Sui Network.

A transaction’s Gas Budget is defined in SUI units and transactions will be successfully executed if:

$$
\text{GasBudget}[\tau]\ \geq\ \text{GasFees}[\tau]
$$

If the Gas Budget does not fulfill this condition — and thus is insufficient to cover a transaction’s gas fees — then the transaction will fail and the entire Gas Budget will be charged. 

> The minimum Gas Budget is 2000 MIST. This ensures validators can be compensated with at least 2000 MIST even if the Gas Budget is mis-specified and transaction aborts. Additionally, this protects the Sui Network from being spammed with a large number of transactions with minimal gas budgets. The maximum Gas Budget is 50 billion MIST or 50 SUI. This protects the network against overflow of internal multiplications and gas limits for denial of service attack

Reviewer 2: ```suggestion
The minimum Gas Budget is 2,000 MIST. This ensures validators can be compensated with at least 2,000 MIST even if the Gas Budget is mis-specified and transaction aborts. Additionally, this protects the Sui Network from being spammed with a large number of transactions with minimal gas budgets. The maximum Gas Budget is 50 billion MIST or 50 SUI. This protects the network against overflow of internal multiplications and gas limits for denial of service attack
```

Reviewer 2: Why capitalize Gas Budget?

> Finally, Sui’s [Storage mechanics](https://docs.sui.io/learn/tokenomics/storage-fund#storage-fund-rewards) provide storage fee rebates whenever a transaction deletes previously-stored objects. Hence, the net fees that a user pays equals Gas Fees minus the rebates associated with data deletion:

Reviewer: I would suggest moving this part up into "Gas units" (since storage rebate affects the total tx cost calculation explained in that section, and is not related to gas budgets).

$$
\text{NetGasFees}[\tau]\ =\ \text{GasFees}[\tau]\ -\text{StorageRebate}[\tau]
$$

> Since the Gas Budget applies to the totality of Gas Fees, it will often be the case that a transaction will only go through if the Gas Budget is considerably higher than the Net Gas Fees that a user ultimately pays.

Reviewer 2: Why cappy Gas Fee?
